pkg/generators/models/models.go

```diff
@@ -599,77 +624,77 @@ func structPropsFromRef(ref *openapi3.SchemaRef) (specs []PropSpec, imports map[
 
 		spec.JSONTags = jsonTags
 
-		err := fillValidationSpecForProperty(prop, &spec, imports)
+		err := fillValidationSpec(prop, &spec.ValidationSpec, spec.GoType, imports)
 		if err != nil {
 			return nil, nil, fmt.Errorf("invalid %s property %q: %w", ExtensionPatternError, name, err)
 		}
+
+		resolveOZZORequiredValidations(&spec)
 		specs = append(specs, spec)
 	}
 
 	return specs, imports, err
 }
 
-// fillValidationSpecForProperty sets validation rules for the property descriptor out of format and
-// bounds in the spec.
-// Returns a list of unique imports in order to implement the validation.
-func fillValidationSpecForProperty(ref *openapi3.SchemaRef, spec *PropSpec, imports map[string]string) (err error) {
-	// Only add validation fields that are part of the property and not the referenced schema.
-	if ref.Ref != "" {
-		valSpec := ValidationSpec{}
-		err = fillValidationSpec(ref, &valSpec, spec.GoType, spec.IsRequired, make(map[string]string))
-		if err != nil {
-			return err
-		}
-		// Enable to force validation on referenced types.
-		spec.NeedsValidation = valSpec.NeedsValidation
-		spec.IsRequiredInValidation = !valSpec.IsNullable && spec.IsRequired
-		return nil
-	}
-
-	return fillValidationSpec(ref, &spec.ValidationSpec, spec.GoType, spec.IsRequired, imports)
+// resolveOZZORequiredValidations resolves the OZZO validation rules that are derived from the required, nullable and other
+// properties that affect the validation rules. Example: OZZO min length validation for strings consider the empty string
+// valid, so we need to add the OZZO Required validation to catch that.
+func resolveOZZORequiredValidations(spec *PropSpec) {
```

Comment:
I also have the same concern, it seems counter intuitive to have to add the null case as well for nullable enums, but I would prefer the generator is consistent with the spec. These are the cases as I understood them:

- nullable: true and null value in the list -> make the field nullable.
- nullable: true and no null value in the list -> This is the case they say the enum is not nullable.
- nullable: false and null value in the list -> make the field nullable, just like the first case.
- nullable: false and no null value in the list -> This is the default case where it is just a regular enum.

So this PR should be update to follow this. Let me know your thoughts.

Reply:
I don't really love this consequence
> nullable: true and no null value in the list -> make the field nullable (pointer), but a nil value is not correct.
it seems weird to allow the data model to represent an invalid value. I would rather the Go model not allow invalid data, if it is possible. This feels like it results in worse Go code that is either harder to work with or could quietly allow invalid data. 

I think it is reasonable that we need to take the sum of the whole spec and generate the best code to represent that. The `nullable: true` expands the `type` field, but the `enum` values then restrict it. If we can clearly see that `null` is not allowed, then we shouldn't allow the resulting model to represent it. 

I don't think this case agrees with their clarification. 
> nullable: false and null value in the list -> this is equivalent to the regular case, the null value in the list should be ignored.

They state several times that `nullable: false` isn't actually making a statement about validation or the type, rather that the `type` field is not modified in any way.  From [their proposal](https://github.com/OAI/OpenAPI-Specification/blob/main/proposals/2019-10-31-Clarify-Nullable.md#proposed-solution)
> A false value leaves the specified or default type unmodified. 

They then [further explain](https://github.com/OAI/OpenAPI-Specification/blob/main/proposals/2019-10-31-Clarify-Nullable.md#detailed-design) that 

> nullable is only meaningful if its value is true.

and 

> It is now clear that nullable: false, whether explicit or by default, does not prohibit null values. Consistent with JSON Schema, an empty object allows all values, including null.

This implies that `nullable: false` does not create any kind of constraint on the type, the inclusion of `null` in the `enum` list is valid _if_ the value is untyped, 

```yaml
nullable: false
enum: ['abc', 1, null]
```
is theoretically allowed, but I don't think we actually support these untyped fields currently. I guess they would have to be `interface{}`.

I think this would however be true

> type: is non-empty, nullable: false, and null value in the list -> produce a run-time error because the spec is invalid

for example

```yaml
type: string
nullable: false
enum: [null, 'a', 'b'] 
```
the `type: string` implies that the value is not nullable (because null is not a string), so the enum values are conflicting. The `nullable: false` actually a no-op in this case.

This stuff becomes a bit dizzying. 

Having said all of that, I think I agree with your table, but I would propose that we only create ptr types if it is `nullable: true` and `null` is in the list.

Reply:
I think the generator should respect the intention of the API designer, so if they want a nullable field, then the generator should generate a pointer field. This allows for the generated code to not change the type later when they add a `null` value, or remove it. That change would only affect the validation, but not the code that interacts with the generated types.

Reply:
I agree it feels weird to declare a nullable enum that can't be null, but the second part is derived from the Validation definition, while it being nullable is in the Type definition.

Reply:
I think there is a bit of a gap here between the intention of the designer and the implementation, since OpenAPI is language agnostic. Not all languages have pointers, for example, and the `nullable` would simply result in in a change to the validation in that code base. If the author of the spec is working in a different server/client language, they might have different intentions. We also have the issue that they could be using a non-compliant generator, so their intent and their spec might be slightly different. 

I am not so concerned that the type might change based on adding or removing the `null` because. For example, the more concerning case is the breaking change when it is removed. This would change a ptr field into a plan field (likely a string). 

Let's assume we leave it as a pointer. In that case, they likely already have `model.enum_field = nil` in some locations and will still need to change these values because they are invalid. `grep` _should_ find them, but if we change it to a plain string, then the go linter will find exactly every case that is broken. This might require some more work, but the code will be more accurate and can accidentally slip out an invalid value. I find that helpful. 

I am going to ping @megaflo to get another opinion on it. I definitely see both sides and having a third voice would push it one way or the other.

Reply:
These are the two cases I see as designer intentions:

1. You want to create a client to consume an external API. In this case I would say it doesn't really matter what path we want to follow, as we can't tell what was the designer intention, as you said, we don't even know if they are thinking in pointers.

2. You are defining the API of your service, and you are using the generated Models all around your business code. You want your Enum to be nullable and do not accept the null as valid input value from the REST layer, but you need it to be a pointer in your business logic to play nice with the rest of your code.

Reply:
Another point, I just checked what is the openapi-generator doing and it doesn't understand zero value nor null in the list, so the generated code will fail to validate a correct value. It does generate a pointer if you tell it to be a nullable prop.
`openapi-generator generate -g go-server --skip-validate-spec  -i api.yaml -o generated`

Reply:
@injeniero This looks awesome, thank you for your work and this PR.

This discussion is really interesting. I was not aware how much extra effort has been put into "fixing" the definition of `nullable` in the specs. And I guess this exchange highlights how difficult it is to settle on an interpretation that works for all implementations.

On the question on how the generated code should look like for nullable enums without null value, I think I lean towards the no pointer solution. My reasoning is, that the generator generally aims to produce Go types that model the specs (or the "allowed values") as closely as possible. The allows much of the validation work to be done by the language itself (meaning the type system and runtime type machinery). In the case of enum properties, that cannot be null, this means that we probably want to prevent `nil` values on the type level.

I think one important consequence of this (and I know that not everyone may agree with me on this one), is that the generated code is "just" the API model and may not always be the best candidate for the domain model. So if the internal business logic of a program requires different values than the API allows, then a custom set of handcrafted structs that is then converted to and from the API model is probably the way to go.

Reply:
Thanks for taking the time to go over it!

It is settle then, I will update the PR to reflect this decision.

Reply:
Changes pushed

Reply:
hi @injeniero if this is something you would really want to support, I think it would be reasonable to consider adding a CLI flag to toggle it. But we can merge this first and add that as a separate feature later.